python/ffsim/states/sample_slater.py

```diff
@@ -211,3 +403,96 @@ def _generate_probs(rdm: np.ndarray, sample: set[int], norb: int) -> np.ndarray:
         indices = list(sample | {orb})
         probs[i] = np.linalg.det(rdm[indices][:, indices]).real
     return probs / np.sum(probs)
+
+
+class _FastProjectionSampler:
+    """Projection-based fast sampler core with cached precomputation.
+
+    Given an orthonormal orbital matrix for a projector 1-RDM, the sampler
+    precomputes the normal vectors needed for the Gaussian-elimination sampling
+    routine and exposes a ``sample`` method that generates integer-encoded
+    bitstrings using a provided RNG.
+    """
+
+    def __init__(self, orbitals: np.ndarray):
+        """Initialize the sampler with occupied orbitals from a projector 1-RDM.
+
+        Precomputes dual normal vectors via reduced QR factorization and
+        triangular solves for reuse during sampling.
+
+        Args:
+            orbitals: A 2D array whose columns are orthonormal orbitals spanning
+                the occupied subspace. Shape ``(norb, nelec)``.
+
+        Raises:
+            ValueError: If the input is not 2D, has zero occupied orbitals, or is
+                ill-conditioned for the QR-based precomputation.
+        """
+        if orbitals.ndim != 2:
+            raise ValueError("Projector orbitals must be a 2D array.")
+        self._norb, self._nelec = orbitals.shape
+        if self._nelec == 0:
+            raise ValueError("Fast sampler requires at least one occupied orbital.")
+        # Precompute normal vectors h_t from QR
+        q, r = np.linalg.qr(orbitals, mode="reduced")
+        diag = np.abs(np.diag(r))
+        tol = 1e-12
+        if np.any(diag <= tol):
+            raise ValueError(
+                "Projector factorization is ill-conditioned for fast sampling."
+            )
+        inv_rt = scipy.linalg.solve_triangular(
+            r.T, np.eye(self._nelec), lower=True, check_finite=False
+        )
+        self._normals = q @ inv_rt
+
+    @property
+    def norb(self) -> int:
+        return self._norb
+
+    @property
+    def nelec(self) -> int:
+        return self._nelec
+
+    def sample(self, rng: np.random.Generator) -> int:
+        """Draw one configuration using the cached normals and supplied RNG.
+
+        Permutes the precomputed normal vectors, then iteratively picks an
+        orbital with probability proportional to the squared magnitude of the
+        leading normal, performs a Gaussian-elimination pivot to update the
+        remaining normals, and repeats until all electrons are placed.
+
+        Args:
+            rng: NumPy Generator to use for randomness.
+
+        Returns:
+            Integer-encoded bitstring for the sampled configuration.
+        """
+        perm = rng.permutation(self._nelec)
+        h = self._normals[:, perm].copy()
+        selected: list[int] = []
+        active = self._nelec
+        tol = 1e-12
+        # Iterate until all electrons are placed
+        while active:
+            h0 = h[:, 0]
```

Comment:
If you can think of descriptive names for `h` and `h0`, that would be preferable to single-letter names.